src/read/xz.rs

```diff
@@ -0,0 +1,330 @@
+use crc32fast::Hasher;
+use lzma_rs::decompress::raw::Lzma2Decoder;
+use std::{
+    collections::VecDeque,
+    io::{BufRead, BufReader, Error, Read, Result, Write},
+};
+
+#[derive(Debug)]
+pub struct XzDecoder<R> {
+    compressed_reader: BufReader<R>,
+    stream_size: usize,
+    buf: VecDeque<u8>,
+    check_size: usize,
+    records: Vec<(usize, usize)>,
+    flags: [u8; 2],
+}
+
+impl<R: Read> XzDecoder<R> {
+    pub fn new(inner: R) -> Self {
+        XzDecoder {
+            compressed_reader: BufReader::new(inner),
+            stream_size: 0,
+            buf: VecDeque::new(),
+            check_size: 0,
+            records: vec![],
+            flags: [0, 0],
+        }
+    }
+}
+
+struct CountReader<'a, R: BufRead> {
+    inner: &'a mut R,
+    count: &'a mut usize,
+}
+
+impl<R: BufRead> Read for CountReader<'_, R> {
+    fn read(&mut self, buf: &mut [u8]) -> Result<usize> {
+        let count = self.inner.read(buf)?;
+        *self.count += count;
+        Ok(count)
+    }
+}
+
+impl<R: BufRead> BufRead for CountReader<'_, R> {
+    fn fill_buf(&mut self) -> Result<&[u8]> {
+        self.inner.fill_buf()
+    }
+
+    fn consume(&mut self, amt: usize) {
+        self.inner.consume(amt);
+        *self.count += amt;
+    }
+}
+
+struct BufWriter<'a> {
+    inner: &'a mut [u8],
+    written: &'a mut usize,
+    total: &'a mut usize,
+    rest: &'a mut VecDeque<u8>,
+}
+
+impl<'a> Write for BufWriter<'a> {
+    fn write(&mut self, buf: &[u8]) -> Result<usize> {
+        if self.inner.len() > *self.written {
+            let len = std::cmp::min(buf.len(), self.inner.len() - *self.written);
+            self.inner[*self.written..*self.written + len].copy_from_slice(&buf[..len]);
+            *self.written += len;
+            *self.total += buf.len();
+            Ok(len)
+        } else {
+            self.rest.extend(buf.iter().copied());
+            *self.total += buf.len();
+            Ok(buf.len())
+        }
+    }
+
+    fn flush(&mut self) -> Result<()> {
+        Ok(())
+    }
+}
+
+fn get_multibyte<R: Read>(input: &mut R, mut hasher: Option<&mut Hasher>) -> Result<u64> {
+    let mut result = 0;
+    for i in 0..9 {
+        let mut b = [0u8; 1];
+        input.read_exact(&mut b)?;
+        if let Some(hasher) = hasher.as_mut() {
+            hasher.update(&b);
+        }
+        let b = b[0];
+        result ^= ((b & 0x7F) as u64) << (i * 7);
+        if (b & 0x80) == 0 {
+            return Ok(result);
+        }
+    }
+    Err(Error::new(
+        std::io::ErrorKind::InvalidData,
+        "Invalid multi-byte encoding",
+    ))
+}
+
+impl<R: Read> Read for XzDecoder<R> {
+    fn read(&mut self, buf: &mut [u8]) -> Result<usize> {
+        if !self.buf.is_empty() {
+            let len = std::cmp::min(buf.len(), self.buf.len());
+            for (i, b) in self.buf.drain(..len).enumerate() {
```

Comment:
Combine these into one big operation using `copy_from_slice` and `VecDeque::as_slices`, for performance reasons mainly involving vectorization.